datafusion/expr/src/logical_plan/plan.rs

```diff
@@ -1013,92 +1046,157 @@ impl LogicalPlan {
                     logical_optimization_succeeded: e.logical_optimization_succeeded,
                 }))
             }
             LogicalPlan::Prepare(Prepare {
                 name, data_types, ..
-            }) => Ok(LogicalPlan::Prepare(Prepare {
-                name: name.clone(),
-                data_types: data_types.clone(),
-                input: Arc::new(inputs.swap_remove(0)),
-            })),
+            }) => {
+                self.check_no_expressions(expr);
+                let input = self.only_input(inputs);
+                Ok(LogicalPlan::Prepare(Prepare {
+                    name: name.clone(),
+                    data_types: data_types.clone(),
+                    input: Arc::new(input),
+                }))
+            }
             LogicalPlan::TableScan(ts) => {
-                assert!(inputs.is_empty(), "{self:?}  should have no inputs");
+                self.check_no_inputs(inputs);
                 Ok(LogicalPlan::TableScan(TableScan {
                     filters: expr,
                     ..ts.clone()
                 }))
             }
             LogicalPlan::EmptyRelation(_)
             | LogicalPlan::Ddl(_)
-            | LogicalPlan::Statement(_) => {
+            | LogicalPlan::Statement(_)
+            | LogicalPlan::DescribeTable(_) => {
                 // All of these plan types have no inputs / exprs so should not be called
-                assert!(expr.is_empty(), "{self:?} should have no exprs");
-                assert!(inputs.is_empty(), "{self:?}  should have no inputs");
+                self.check_no_expressions(expr);
+                self.check_no_inputs(inputs);
                 Ok(self.clone())
             }
-            LogicalPlan::DescribeTable(_) => Ok(self.clone()),
             LogicalPlan::Unnest(Unnest {
                 exec_columns: columns,
                 options,
                 ..
             }) => {
+                self.check_no_expressions(expr);
+                let input = self.only_input(inputs);
                 // Update schema with unnested column type.
-                let input = inputs.swap_remove(0);
                 let new_plan =
                     unnest_with_options(input, columns.clone(), options.clone())?;
                 Ok(new_plan)
             }
         }
     }
+
+    #[allow(clippy::needless_pass_by_value)] // expr is moved intentionally to ensure it's not used again
+    fn check_no_expressions(&self, expr: Vec<Expr>) {
+        assert!(
+            expr.is_empty(),
+            "{self:?} should have no exprs, got {:?}",
+            expr
+        );
+    }
+
+    #[allow(clippy::needless_pass_by_value)] // inputs is moved intentionally to ensure it's not used again
+    fn check_no_inputs(&self, inputs: Vec<LogicalPlan>) {
+        assert!(
+            inputs.is_empty(),
+            "{self:?} should have no inputs, got: {:?}",
+            inputs
+        );
+    }
+
+    fn only_expr(&self, mut expr: Vec<Expr>) -> Expr {
```

Comment:
I doubt whether this is helpful, a little over abstraction IMO 🤔

Reply:
i had this initially inline but then the check code would take more space, and it was making it harder to see the important part of the logic (obviously this is subjective)
another subjective bonus is there is type system help -- i cannot check there are no expressions or that there is only one expression and still use expressions vector to pull something later from it.

if you would prefer to have these functions inlined back, just let me know, happy to do so

Reply:
Maybe we could add some comments explaining the rationale as a way to make the reason for the extra abstraction clear

Reply:
done!